### ISSUE TITLE
assertion failure in Element.as(JavaScriptObject o)

### ISSUE THREAD

**Issue by bill-reck** [gwtproject/gwt#9610]
**GWT version: 2.8.2** 
**Browser (with version): firefox 24.1.0** 
**Operating System: linux/slackware** 

---

##### Description

I get a assertion failure from Element.as(JavaScriptObject o) when I drag over a CellTree with assertions enabled in production mode with firefox.
It doesn't happen with chrome.  I haven't tried the fail case with any other browsers or platforms.

I am overriding AbstractCell.onBrowserEvent in my CellTree but my classes don't appear in the stack trace so I'm assuming its a GWT problem.

stack trace:

com.google.gwt.core.shared.SerializableThrowable: null
	at Unknown.java_lang_Throwable_Throwable__V(http://crow) ~[na:na]
	at Unknown.java_lang_AssertionError_AssertionError__V(http://crow) ~[na:na]
	at Unknown.com_google_gwt_dom_client_Element_as__Lcom_google_gwt_core_client_JavaScriptObject_2Lcom_google_gwt_dom_client_Element_2(http://crow) ~[na:na]
	at Unknown.com_google_gwt_user_cellview_client_CellTreeNodeView_$fireEventToCell__Lcom_google_gwt_user_cellview_client_CellTreeNodeView_2Lcom_google_gwt_dom_client_NativeEvent_2V(http://crow) ~[na:na]
	at Unknown.com_google_gwt_user_cellview_client_CellTree_onBrowserEvent__Lcom_google_gwt_user_client_Event_2V(http://crow) ~[na:na]
	at Unknown.com_google_gwt_user_client_DOM_dispatchEvent__Lcom_google_gwt_user_client_Event_2Lcom_google_gwt_dom_client_Element_2Lcom_google_gwt_user_client_EventListener_2V(http://crow) ~[na:na]
	at Unknown.com_google_gwt_user_client_impl_DOMImplStandard_dispatchEvent__Lcom_google_gwt_user_client_Event_2V(http://crow) ~[na:na]
	at Unknown.com_google_gwt_user_client_impl_DOMImplStandard_dispatchDragEvent__Lcom_google_gwt_user_client_Event_2V(http://crow) ~[na:na]
	at Unknown.com_google_gwt_core_client_impl_Impl_apply__Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2(http://crow) ~[na:na]
	at Unknown.com_google_gwt_core_client_impl_Impl_entry0__Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2(http://crow) ~[na:na]
	at Unknown.com_google_gwt_core_client_impl_Impl_entry__Lcom_google_gwt_core_client_JavaScriptObject_2Lcom_google_gwt_core_client_JavaScriptObject_2/<(http://crow) ~[na:na]
	at Unknown.anonymous(Unknown) ~[na:na]

##### Steps to reproduce
-- enable assertions in production mode
-- click on a CellTree
-- drag the mouse until its just out of the CellTree
-- assertion failure

**Comment by bill-reck** [comment]
Thanks very much for the prompt response and your time.  I'll look into it
further along the lines of your suggestions.

**Comment by bill-reck** [comment]
You're right -- it is an issue related to firefox.  I tried the fail case on firefox 59.0.1/Windows 10 and got the same error as with firefox 24 on linux.

I've patched CellTreeNodeView with a predicate that bypasses the assert and logs the information you requested.  The patch works with firefox and chrome.  I'm not really sure if its the correct logic.  Please see the attachments.

[CellTreeNodeView_patched.java.txt](https://github.com/gwtproject/gwt/files/1922334/CellTreeNodeView_patched.java.txt)

[drag_assertion_failure_with_firefox_log.txt](https://github.com/gwtproject/gwt/files/1922335/drag_assertion_failure_with_firefox_log.txt)

**Comment by bill-reck** [comment]
Didn't mean to close it

**Comment by tbroyer** [comment]
OK, so, in this specific case, it's a bug in Firefox (`dragenter` and `dragleave` aren't supposed to have non-element targets; see https://html.spec.whatwg.org/multipage/dnd.html#immediate-user-selection which explicitly says “Only elements can be selected by the user; other nodes must not be made available as drop targets.” and https://html.spec.whatwg.org/multipage/dnd.html#dndevents which indicates the targets for each event type); but `dragstart`, `drag` and `dragend` can actually be dispatched at text nodes (see https://html.spec.whatwg.org/multipage/dnd.html#source-node) and GWT needs to account for this.

A quick fix would be to change the `Element.as()` to a `Node.as()` in `CellTreeNodeView`. A proper fix would however also eliminate the `event.getEventTarget().cast()` in `CellTree`.
